File: ch09/random_select.java

import java.util.ArrayList;
import java.util.Random;

class Select {

  private static int partition(ArrayList<Integer> v, int left, int right) {
    int i = left - 1;
    for (int j = left; j != right; j += 1) {
      if (v.get(j) < v.get(right)) {
        i += 1;
        Main.exchange(v, i, j);
      }
    }
    i += 1;
    Main.exchange(v, i, right);
    return i;
  }

  private static int random_partition(ArrayList<Integer> v,
      int left, int right) {
    int p = new Random().nextInt(right - left + 1) + left;
    Main.exchange(v, p, right);
    return partition(v, left, right);
  }

  private static Integer random_select_b(int index,
      ArrayList<Integer> v, int left, int right) {

    class Help {
      Integer rs_help(int l, int r) {
        if (l < r) {
          int p = random_partition(v, l, r);
          if (index < p) {
            rs_help(l, p - 1);
          } else if (index > p) {
            rs_help(p + 1, r);
          } //else {
            return v.get(index);
          //}
        } else {
          return v.get(index);
        }
      }
    }

    return new Help().rs_help(left, right);
  }

  public static Integer random_select(int index,
      ArrayList<Integer> v, int left, int right) {

    if (left < 0) {
      throw new IllegalArgumentException("random_select: left");
    } else if (right >= v.size()) {
      throw new IllegalArgumentException("random_select: right");
    } else if (index < left || index > right) {
      throw new IllegalArgumentException("random_select: index");
    } else {
      return random_select_b(index, v, left, right);
    }
  }
}

//--------------------------------------

class Main {

  static void insertion_sort(ArrayList<Integer> v) {
    int n = v.size();
    if (n > 0) {
      for (int i = 1; i != n; i += 1) {
        for (int p = i - 1, q = i;
            p >= 0 && v.get(p) > v.get(q); p -= 1, q -= 1) {
          exchange(v, p, q);
        }
      }
    }
  }

  static <T> void exchange(ArrayList<T> v, int i, int j) {
    T temp = v.get(i);
    v.set(i, v.get(j));
    v.set(j, temp);
  }

  static void setRandomArray(
      ArrayList<Integer> v, int size, int min, int max) {
    Random rand = new Random();
    int bound = max - min;
    for (int i = 0; i < size; i += 1) {
      v.add(rand.nextInt(bound) + min);
    }
  }

  static <T> void printArray(ArrayList<T> v) {
    System.out.print("(ArrayList #");
    System.out.print(v.size() + " [");
    for (int i = 0; i < v.size(); i += 1) {
      System.out.print(v.get(i) + " ");
    }
    System.out.println("])");
  }

  public static void main(String[] args) {

    ArrayList<Integer> A = new ArrayList<>();
    setRandomArray(A, 100, 0, 100);
    printArray(A);
    System.out.println(Select.random_select(95, A, 0, 99));
    insertion_sort(A);
    printArray(A);
  }
}

File: ch09/random_select.cpp
#include <iostream>
#include <vector>
#include <random>
#include <functional>
#include <stdexcept>

using std::cout;
using std::vector;


template<typename BaseData>
void print_vector(vector<BaseData>&);
template<typename BaseData>
void exchange(vector<BaseData>&, int, int);
long random(long, long);


template<typename BaseData>
int partition(vector<BaseData>& v, int left, int right) {
    int i = left - 1;
    for (int j = left; j != right; j += 1) {
        if (v[j] < v[right]) {
            i += 1;
            exchange(v, i, j);
        }
    }
    i += 1;
    exchange(v, i, right);
    return i;
}

template<typename BaseData>
int random_partition(vector<BaseData>& v, int left, int right) {
    int p = random(left, right + 1);
    exchange(v, p, right);
    return partition<BaseData>(v, left, right);
}


long random_select_b(int index,
        vector<long>& v, int left, int right) {

    std::function<long (int, int)> RS_help = [&](int l, int r) {
        if (l < r) {
            int p = random_partition(v, l, r);
            if (index < p) {
                RS_help(l, p - 1);
            } else if (index > p) {
                RS_help(p + 1, r);
            } else {
                return v[index];
            }
            //return v[index];  // damn clang++, why need this one?
        } else {
            return v[index];
        }
    };

    return RS_help(left, right);
}

long random_select(int index,
        vector<long>& v, int left, int right) {
    if (left < 0) {
        throw std::invalid_argument("left");
    } else if (right >= (int)v.size()) {
        throw std::length_error("right");
    } else if (left > index || index > right) {
        throw std::out_of_range("index");
    } else {
        return random_select_b(index, v, left, right);
    }
}


//--------------------------------------

template<typename BaseData>
void exchange(vector<BaseData>& v, int i, int j) {
    BaseData temp = v[i];
    v[i] = v[j];
    v[j] = temp;
}

template<typename BaseData>
void quick_sort(vector<BaseData>& v) {
    auto partition = [&](int l, int r) {
        int k = l - 1;
        for (int i = l; i != r; i += 1) {
            if (v[i] < v[r]) {
                k += 1;
                exchange(v, i, k);
            }
        }
        k += 1;
        exchange(v, k, r);
        return k;
    };

    std::function<void (int, int)> quick_sort_asc = [&](int l, int r) {
        if (l < r) {
            int q = partition(l, r);
            quick_sort_asc(l, q - 1);
            quick_sort_asc(q + 1, r);
        }
    };

    quick_sort_asc(0, v.size() - 1);
}

long random(long min, long max) {
    std::random_device rd;
    std::mt19937_64 gen(rd());
    std::uniform_int_distribution<long> dis(min, max - 1);
    return dis(gen);
}

void set_vector(vector<long>& v, int size, long min, long max) {
    for (int i = 0; i < size; i = i+1) {
        v.push_back(random(min, max));
    }
}

template<typename BaseData>
void print_vector(vector<BaseData>& v) {
    cout << "(vector #";
    cout << v.size() << " [";
    for (int i = 0; i < (int)v.size(); i = i+1) {
        cout << v[i] << " ";
    }
    cout << "])\n";
}

int main() {
    vector<long> A;
    set_vector(A, 100, 0, 100);
    print_vector(A);
    cout << random_select(97, A, 0, 99) << "\n";
    quick_sort(A);
    print_vector(A);
    return 0;
}
